scripts/plugins/rum.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * log RUM if part of the sample.
+ * @param {string} checkpoint identifies the checkpoint in funnel
+ * @param {Object} data additional data for RUM sample
+ */
+function sampleRUM(checkpoint, data = {}) {
+  sampleRUM.defer = sampleRUM.defer || [];
+  const defer = (fnname) => {
+    sampleRUM[fnname] = sampleRUM[fnname]
+      || ((...args) => sampleRUM.defer.push({ fnname, args }));
+  };
+  sampleRUM.drain = sampleRUM.drain
+    || ((dfnname, fn) => {
+      sampleRUM[dfnname] = fn;
+      sampleRUM.defer
+        .filter(({ fnname }) => dfnname === fnname)
+        .forEach(({ fnname, args }) => sampleRUM[fnname](...args));
+    });
+  sampleRUM.on = (chkpnt, fn) => { sampleRUM.cases[chkpnt] = fn; };
+  defer('observe');
+  defer('cwv');
+  try {
+    window.hlx = window.hlx || {};
+    if (!window.hlx.rum) {
+      const usp = new URLSearchParams(window.location.search);
+      const weight = (usp.get('rum') === 'on') ? 1 : 100; // with parameter, weight is 1. Defaults to 100.
+      // eslint-disable-next-line no-bitwise
+      const hashCode = (s) => s.split('').reduce((a, b) => (((a << 5) - a) + b.charCodeAt(0)) | 0, 0);
+      const id = `${hashCode(window.location.href)}-${new Date().getTime()}-${Math.random().toString(16).substr(2, 14)}`;
+      const random = Math.random();
+      const isSelected = (random * weight < 1);
+      // eslint-disable-next-line object-curly-newline
+      window.hlx.rum = { weight, id, random, isSelected, sampleRUM };
+    }
+    const { weight, id } = window.hlx.rum;
+    if (window.hlx && window.hlx.rum && window.hlx.rum.isSelected) {
+      const sendPing = (pdata = data) => {
+        // eslint-disable-next-line object-curly-newline, max-len, no-use-before-define
+        const body = JSON.stringify({ weight, id, referer: window.location.href, generation: window.hlx.RUM_GENERATION, checkpoint, ...data });
+        const url = `https://rum.hlx.page/.rum/${weight}`;
+        // eslint-disable-next-line no-unused-expressions
+        navigator.sendBeacon(url, body);
+        // eslint-disable-next-line no-console
+        console.debug(`ping:${checkpoint}`, pdata);
+      };
+      sampleRUM.cases = sampleRUM.cases || {
+        cwv: () => sampleRUM.cwv(data) || true,
+        lazy: () => {
+          // use classic script to avoid CORS issues
+          const script = document.createElement('script');
+          script.src = 'https://rum.hlx.page/.rum/@adobe/helix-rum-enhancer@^1/src/index.js';
```

Comment:
The code loaded behind the scenes assumes that `sampleRUM` is a global object. Your changes break that contract, rendering the advanced functionality useless.

Reply:
@trieloff You mean that it is assuming it's exposed on `window.hlx.rum`, right? This is still the case in the plugin. The instrumentation is just isolated in the plugin and decoupled from `scripts.js`, but I haven't changed existing APIs as far as I can tell.

Reply:
<img width="1470" alt="Screen Shot 2022-10-24 at 15 52 08" src="https://user-images.githubusercontent.com/39613/197636801-f6e53765-93f7-41a3-8563-febcd8c3184e.png">

The `load` checkpoint is never sent from your refactored script. See #139 (which will conflict) for the correct checkpoint sequence.

Reply:
@trieloff Thanks for sharing, that helps set clear expectations. Are there clear expectations for each event? Do we have those documented somewhere?
For instance, does the load event need to be triggered on the page load? or is it meant to reflect a specific moment in the loading phase?

Reply:
There are various ways I can fix the missing `load` event depending on the exact expectation.

Since the RUM plugin is loaded async, essentially the load event would have fired by the time any code in it is executed. So, in theory, I can bring it back into the main script and lose the separation of concern if we care about the exact page load event.

But I could also just fire it whenever the rum plugin has loaded, if what we care is when the main JS logic is loaded.

Or I could defer to when the content becomes visible (waitForLCP), if that is what we really care about.

It really depends what we use that specific event for in the end in the analytics reports

Reply:
For instance https://github.com/adobe/helix-project-boilerplate/pull/135/commits/0d2bab2646aaf325699e10707f36c6034da3e8cc adds the event back, and we essentially now have:
- `top` is fired when the plugin is "parsed"
- `load` is fired when the plugin is "initialized"

Reply:
`top` should be fired as soon as possible. The goal is to catch any hint of the page being interacted with at all.
`load` is the https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event plain and simple.

Reply:
@trieloff I moved the rum plugin back into lib-franklin so we can trigger on the DOM load event